app/api/ada/adaTransactions/adaNewTransactions.js

```diff
@@ -46,107 +28,149 @@ const protocolMagic = CONFIG.network.protocolMagic;
 /** Calculate the transaction fee without actually sending the transaction */
 export async function getAdaTransactionFee(
   receiver: string,
-  amount: string
+  amount: string,
+  changeAdaAddr: ?AdaAddress,
+  possibleInputAddresses: Array<AdaAddress>,
+  addressesToUtxos: Array<string> => Promise<Array<UTXO>>
 ): Promise<AdaFeeEstimateResponse> {
-  const { txBuilder } = await newAdaUnsignedTx(receiver, amount);
-  const fee = txBuilder.estimate_fee(RustModule.Wallet.LinearFeeAlgorithm.default());
-  return { fee: fee.to_str() };
+  const { txBuilder } = await newAdaUnsignedTx(
+    receiver,
+    amount,
+    changeAdaAddr,
+    possibleInputAddresses,
+    addressesToUtxos
+  );
+  const fee = txBuilder.get_balance_without_fees().value();
+  return { fee };
 }
 
-export async function newAdaUnsignedTx(
+export async function sendAllUnsignedTxFromUtxo(
   receiver: string,
-  amount: string,
-): Promise<UnsignedTxResponse> {
-  const changeAdaAddr: AdaAddress = await popBip44Address('Internal');
-  const changeAddr = RustModule.Wallet.Address.from_base58(changeAdaAddr.cadId);
-  const outputPolicy = RustModule.Wallet.OutputPolicy.change_to_one_address(changeAddr);
+  allUtxos: Array<UTXO>,
+): Promise<UnsignedTxFromUtxoResponse> {
+  const totalBalance = allUtxos
+    .map(utxo => new BigNumber(utxo.amount))
+    .reduce(
+      (acc, amount) => acc.plus(amount),
+      new BigNumber(0)
+    );
+
   const feeAlgorithm = RustModule.Wallet.LinearFeeAlgorithm.default();
+  // firts build a transaction to see what the cost would be
+  const fakeTxBuilder = new RustModule.Wallet.TransactionBuilder();
+  const inputs = utxoToTxInput(allUtxos);
+  addTxInputs(fakeTxBuilder, inputs);
+  const inputAmount = coinToBigNumber(fakeTxBuilder.get_input_total());
+  addOutput(fakeTxBuilder, receiver, inputAmount.toString());
+  const fee = coinToBigNumber(fakeTxBuilder.estimate_fee(feeAlgorithm));
 
-  const allAdaAddresses = await getAdaAddressesList();
-  const allUtxos = await getAllUTXOsForAddresses(allAdaAddresses.map(addr => addr.cadId));
-  const txInputs = utxoToTxInput(allUtxos);
-  let selectionResult;
-  try {
-    selectionResult = getInputSelection(txInputs, outputPolicy, feeAlgorithm, receiver, amount);
-  } catch (err) {
+  // create a new transaction subtracing the fee from your total UTXO
+  const newAmount = totalBalance.minus(fee);
+  const unsignedTxResponse = await newAdaUnsignedTxFromUtxo(receiver, newAmount, null, allUtxos);
+
+  // sanity check
+  const balance = unsignedTxResponse.txBuilder.get_balance(feeAlgorithm);
+  if (balance.is_negative()) {
+    /**
+     * The balance may be slightly positive. This is because lowering the "amount" to send
```

Comment:
small typo 
`The balance may be slightly positive` ->  `The balance may be slightly negative`

Reply:
I added more comments to explain what this means